json_serializable/lib/src/type_helpers/iterable_helper.dart

```diff
@@ -70,8 +70,10 @@ class IterableHelper extends TypeHelper {
 
     var itemSubVal = context.deserialize(iterableGenericType, closureArg);
 
-    // If `itemSubVal` is the same, then we don't need to do anything fancy
-    if (closureArg == itemSubVal) {
+    // If `itemSubVal` is the same and it's not a Set, then we don't need to do
+    // anything fancy
+    if (closureArg == itemSubVal &&
```

Comment:
@gagoman – this was the one case that was missing. Tough to find unless you do the full set of cases in the `kitchen_sink` test